test/jdk/javax/swing/JButton/bug4323121.java

```diff
@@ -30,96 +30,100 @@
         robot.setAutoDelay(100);
+
         try {
             SwingUtilities.invokeAndWait(() -> {
+                button = new TestButton("gotcha");
+                button.addMouseListener(new MouseAdapter() {
+                    @Override
+                    public void mouseEntered(MouseEvent e) {
+                        if (button.getModel().isArmed()) {
+                            modelArmed = true;
+                        }
+                        mouseEntered.countDown();
+                    }
+                });
+
                 frame = new JFrame("bug4323121");
-                button = new testButton("gotcha");
                 frame.getContentPane().add(button);
+
+                frame.addWindowFocusListener(new WindowAdapter() {
+                    @Override
+                    public void windowGainedFocus(WindowEvent e) {
+                        windowGainedFocus.countDown();
+                    }
+                });
+
                 frame.pack();
                 frame.setLocationRelativeTo(null);
                 frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                 frame.setVisible(true);
             });
+
+            if (!windowGainedFocus.await(1, SECONDS)) {
+                throw new RuntimeException("Window didn't gain focus");
+            }
             robot.waitForIdle();
-            robot.delay(1000);
+
             SwingUtilities.invokeAndWait(() -> {
-                pt = button.getLocationOnScreen();
-                buttonW = button.getSize().width;
-                buttonH = button.getSize().height;
+                Point location = button.getLocationOnScreen();
+                buttonCenter = new Point(location.x + button.getWidth() / 2,
+                                         location.y + button.getHeight() / 2);
             });
-            robot.mouseMove(pt.x + buttonW / 2, pt.y + buttonH / 2);
-            robot.waitForIdle();
-            if (failed) {
-                throw new RuntimeException("Any created button returns " +
-                                    "true for isArmed()");
+
+            robot.mouseMove(buttonCenter.x, buttonCenter.y);
+
+            if (!mouseEntered.await(1, SECONDS)) {
+                throw new RuntimeException("Mouse entered event wasn't received");
+            }
+            if (modelArmed) {
+                throw new RuntimeException("getModel().isArmed() returns true "
+                                           + "for a subclass of JButton");
             }
         } finally {
-                SwingUtilities.invokeAndWait(() -> {
+            SwingUtilities.invokeAndWait(() -> {
                 if (frame != null) {
                     frame.dispose();
                 }
             });
         }
     }
 
-    static class testButton extends JButton implements MouseMotionListener, MouseListener {
-        public testButton(String label) {
+    private static final class TestButton extends JButton {
```

Comment:
There isn't much custom code within TestButton but it is retained as subclass of JButton because of test requirement?

Reply:
The bug description for [JDK-4323121](https://bugs.openjdk.org/browse/JDK-4323121) mentions a subclass of `JButton`.

However, regular `JButton` also reproduces the bug, I verified that the test fails in 1.3.0 if I use `JButton` directly. The fix for JDK-4323121 is outlined in the comments. The fix modified the common handler for mouse events, before the fix the model had been armed in `mouseEntered`; after the fix the model is armed only if `model.isPressed` returns true.

https://github.com/openjdk/jdk/blob/330f2b5a9cad02b8e6882fc6eee996d7792d3de1/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicButtonListener.java#L290-L291

So, the test can be simplified further by using `JButton` instead of a subclass.